### PR TITLE
Updated functools.wraps and functools.update_wrapper to use ParamSpec…

### PR DESCRIPTION
**Pull request by erictraut**
… to preserve signatures of wrapped and wrapper functions.

### CODE REVIEW

**Comment by erictraut** [comment]
I'm not sure how to make the pytest CI script happy in this case. In fact, it's not obvious to me from the output what it's complaining about. Any suggestions? @JelleZijlstra? @srittau?

**Comment by srittau** [comment]
This is a pytype issue. Maybe @rchen152 has some insight.

File: stdlib/functools.pyi
Comment (srittau) [inline stdlib/functools.pyi:58]: What motivated this change? `update_wrapper()` modifies `wrapper` and then returns it. Is it to add the `__wrapped__` attribute or to handle cases where `assigned` adds some attributes to `wrapper`?

Comment (erictraut) [reply stdlib/functools.pyi:58]: Are you asking specifically about the change for `update_wrapper`? This PR was mostly motivated by a desire to fix `wraps`, but I figured I'd fix `update_wrapper` while I was at it.

The existing definition of `wraps` is pretty broken for a couple of reasons.

First, it leaves the signature of the decorated function unmodified. Typically the decorated function has a signature of `(*args, **kwargs)`, which is quite useless. The purpose of `wraps` is to effectively present a different signature for the decorated function. For example, `@wraps(foo)` should apply the signature of `foo` to the decorated function. This tends to wreak havoc with type checking and completion suggestions. It's one of the most upvoted issues we have with pylance. See https://github.com/microsoft/pylance-release/issues/442 for details.

Second, the existing definition doesn't correctly model the fact that a `__wrapped__` attribute is added to the resulting object. With my proposed change, it not only models the presence of a `__wrapped__` attribute but also provides its correct type (as opposed to a less precise `Callable[..., Any]`).

Comment (srittau) [reply stdlib/functools.pyi:58]: It was specifically asking about `update_wrapper()`. I haven't looked in depth at `wraps()`. (It's a few nestings too deep to think about at 9pm.)

Comment (JelleZijlstra) [reply stdlib/functools.pyi:58]: For what it's worth, my experience is that making `wraps()` fully transparent itself causes issues: https://github.com/quora/pyanalyze/issues/82. Many wrapping functions modify the signature of the wrapped function.

Comment (erictraut) [reply stdlib/functools.pyi:58]: @JelleZijlstra, I'm not clear on what you mean. The link you provided talks about the `@mock.patch` decorator, which is quite different from `@functools.wraps`. How would a wrapping function modify the signature of the wrapped function? If you happen to have an example, that would help me understand.

**Comment by rchen152** [comment]
> This is a pytype issue. Maybe @rchen152 has some insight.

Huh, that's a strange error. It usually occurs when some name is used twice in a stub (e.g.,
```
x: int
x: str
```
) but that's clearly not the case here. I'll take a look.

FYI we're not doing any more pytype releases until after the New Year, since most of the team is OOO, so it'll take a little longer than usually to ship a fix :sweat_smile:

**Comment by srittau** [comment]
One thing to try: As we are effectively "throwing away" the wrapper function's signature (except for the `__wrapped__` field) and replacing it by the wrapped function's, we could use a simple `_Wrapper = TypeVar("_Wrapper", bound=Callable[..., Any])` type var. `_P2` and `_R2` are always used together, and `_P2.args` or `_P2.kwargs` is never accessed. This could at least help us get a clearer picture of what is going wrong.

**Comment by AlexWaygood** [comment]
I wonder if inheriting from `Protocol` rather than `Generic` might reduce the primer diff somewhat?

**Comment by srittau** [comment]
The PR certainly seems worthwhile  and correct to me, but we still need pytype and mypy fixes for it to be viable. I suggest we keep it open as "deferred" for now.

File: stdlib/functools.pyi
Comment (JelleZijlstra) [reply stdlib/functools.pyi:58]: `mock.patch` uses `@wraps` internally. The point here is that many decoorators that use `@wraps` actually don't preserve the signature of the "wrapped" function, so changing the annotations to claim that it does may cause problems.

**Comment by rchen152** [comment]
The pytype issue in this PR has been fixed in the latest release (pytype-2022.1.5).

**Comment by erictraut** [comment]
Thanks @rchen152! Looks like we'll need to update the pytype CI tests in typeshed to use the new version.

@srittau, do you have any remaining concerns about the PR?

**Comment by AlexWaygood** [comment]
The `spark` entry in the primer diff concerns me. It's going to be far more common to use `@wraps` with functions than with any other kind of callable, and it seems that, with this patch, mypy no longer recognises that a function will still have a `__name__` attribute even if it's decorated with `@wraps`. I'm guessing that also applies for `__defaults__`, `__kwdefaults__`, and all the other special attributes that exist for functions but not for other callables.

**Comment by srittau** [comment]
The primer output is still a bit concerning, but I think mostly unavoidable with this PR. Any other opinions on that? I am removing the "deferred" label for now.

**Comment by AlexWaygood** [comment]
> The `spark` entry in the primer diff concerns me. 

I have an idea for how we could get rid of this false-positive (and the similar one for `pandas`), *but*:

- My idea is a little complicated. 
- I'm not sure my idea will work.
- I can always try it out in a separate PR after this one is merged, to see what mypy_primer thinks. 

So, please don't block this PR solely based on my concerns :)

**Comment by erictraut** [comment]
Any other thoughts on this PR? 

Incidentally, the proposed change addresses one of the most upvoted [issues for pylance](https://github.com/microsoft/pylance-release/issues/442), which is why I'd really like to see this fix (or something similar) make it into typeshed.

**Comment by JelleZijlstra** [comment]
Let's do it, though I'm still worried this will create false positives for places that use wraps() but change the signature.

**Comment by JelleZijlstra** [comment]
(And I hope I fixed the merge conflict right this time)

**Review by JelleZijlstra — APPROVED** [review]

**Comment by Akuli** [comment]
> positives for places that use `wraps()` but change the signature

These are arguably true positives, as one reason to use `wraps()` is that it copies the signature. A type checker should warn about this IMO.

**Comment by erictraut** [comment]
Is this PR ready to merge, or are we still waiting for additional input?

**Comment by JelleZijlstra** [comment]
The mypy-primer results indicate this change may be disruptive for mypy. I haven't looked at the code in detail yet.

**Comment by DevilXD** [comment]
@erictraut Is usage like this supposed to give errors?

```py
AnyFunc = TypeVar("AnyFunc", bound=Callable[..., object])

def deco(func: AnyFunc) -> AnyFunc :
    @functools.wraps(func)
    def wrapper(*args, **kwargs) -> object:
        ...
    return wrapper
```
```
Argument 1 to "wraps" has incompatible type "AnyFunc"; expected "Callable[[VarArg(<nothing>), KwArg(<nothing>)], <nothing>]"  [arg-type]
```

According to the primer results, this error type pops up in `zulip` 8 times, once in `pytest` and once in `kopf`. I'm not 100% sure, but the usage above looks valid to me. If that's not the case, please correct me.

**Comment by erictraut** [comment]
I think it's correct for this to be flagged as an error. The signature of the function being wrapped (which is the input parameter `func`) is not the necessarily the same as the signature of the resulting wrapped function, so they cannot be represented by the same type variable. This would probably require a `cast(AnyFunc, wrapper)`. 

Both pyright and mypy flag this as an error today, prior to the proposed change in typeshed's `functools.pyi`.

**Comment by erictraut** [comment]
Yeah, that case is problematic because of the way `ParamSpec` captures the `self` as part of the signature. I don't have a good suggestion for how to work around problem. [Here](https://github.com/python/typing/discussions/946) is a discussion that touches on this issue and asks the authors of PEP 612 for guidance. Unfortunately, I don't think the suggestion they provide in that discussion (which is about `@functools.cache`) works for `@functools.wraps`.

Unless someone has a better suggestion, we'll need to decide whether a false positive in this case (where the argument passed to `wraps()` is an instance method) is common enough to outweigh the benefits that this change otherwise provides. I'll just reiterate that the current handling of `wraps` generates a lot of complaints from pyright and pylance users, so I'm motivated to find some fix.

**Comment by erictraut** [comment]
Are there any updates on this PR? I'm unclear on whether there's anything else I need to do to get it merged.

We continue to get bug reports from pyright users about decorators that use `wraps`. Without this fix, I'm not sure what to tell them.

File: stdlib/functools.pyi
Comment (erictraut) [reply stdlib/functools.pyi:65]: I think it could be confusing to rename these `_WrapperP` and `_WrappedP`. `_P1` is used as a type argument in both the `_Wrapped` and `_Wrapper` types in the `wraps` signature. It's used to translate between the wrapped and the wrapper.

You mentioned that you ran into inconsistencies. Can you elaborate?

File: stdlib/functools.pyi
Comment (srittau) [inline stdlib/functools.pyi:102]: `_P2`/`_R2` are the "wrapper" types. I believe this should read:

```suggestion
    __wrapped__: Callable[_P1, _R1]
```

File: stdlib/functools.pyi
Comment (srittau) [inline stdlib/functools.pyi:103]: On the other hand, `update_wrapper()` doesn't change the wrapper's signature (unless I'm missing something):

```python
from functools import update_wrapper

def wrapper(x: int) -> str:
    return str(x)

def wrapped(y: str, z: bool) -> float:
    return 42.0

wrapper2 = update_wrapper(wrapper, wrapped)

assert wrapper2 is wrapper
print(wrapper(13))
print(wrapper("", True))  # TypeError
```

Same for `wraps()`:

```python
from functools import wraps

def my_decorator(f):
    @wraps(f)
    def wrapper() -> int:
        return 42
    return wrapper

@my_decorator
def wrapped(x: int) -> bool:
    return True

print(type(wrapped()))  # -> class 'int'
print(type(wrapped(42)))  # TypeError
```

So this should read:

```suggestion
    def __call__(self, *args: _P2.args, **kwargs: _P2.kwargs) -> _R2: ...
```

**Review by srittau — COMMENTED** [review]
Sorry for the late reply, but I tried to recreate my thoughts from a few months back. I'm also closing the old conversations to make this clearer.

File: stdlib/functools.pyi
Comment (erictraut) [reply stdlib/functools.pyi]: Good catch. Yes, after looking at your examples above, I realize that I had the ParamSpecs swapped in `_Wrapped`. Now that I've swizzled those, your recommendation for renaming the ParamSpecs makes sense.

**Comment by erictraut** [comment]
Looks like this change is triggering a bug in mypy. :(

**Comment by erictraut** [comment]
On the bright side, it looks like it's eliminating the need for a bunch of existing `# type: ignore` comments. :)

**Comment by erictraut** [comment]
The errors reported in homeassistant appear to be due to a bug in mypy as well. :( These type check without problems in pyright.

**Comment by JelleZijlstra** [comment]
Would it be feasible to first change only `update_wrapper` in a separate PR? That feels lower risk, and that way we can start making some progress.

**Comment by erictraut** [comment]
I don't think anyone uses `update_wrapper` in practice, so that would have little or no value. The real need here is for `wraps` to be properly typed because it's used frequently.

**Comment by JelleZijlstra** [comment]
Good news is that I can reproduce the pyspark crash with mypy 0.950, but not with current mypy master, so I think it's been fixed.

**Comment by AlexWaygood** [comment]
The crashes do indeed seem to be fixed on mypy 0.960! Still a lot of primer output elsewhere, though :/

**Comment by srittau** [comment]
Looking at the primer output, one of the main problems seems to be that mypy has a problem with constructs like the following:

```python
def foo(func: Callable[..., Any]):
    @functools.wraps(func)  # Argument 1 to "wraps" has incompatible type "Callable[..., Any]"; expected "Callable[[VarArg(<nothing>), KwArg(<nothing>)], <nothing>]"  [arg-type]
    def inner(*args, **kwargs):
        ...
```

**Comment by kkirsche** [comment]
> What is the way forward on this issue? It continues to cause pain for pyright users. Mypy now has full support for `ParamSpec`. Is there some other mypy change or bug fix that needs to be implemented?

Following up on this, there has been commit activity since this was asked, but this question still needs to be answered. Alex added the "needs decision" label on Aug 20, 2022, so I was wondering if any concrete steps can move this forward toward a resolution one way or another? Thanks all

**Comment by srittau** [comment]
I'm fine with merging this. The primer output doesn't look great, but not too bad. A few type ignores are okay, until the remaining mypy problems are sorted out.

Unless someone objects, I will do so in the next few days.

**Pull request merged**
